cephfs: expose snapshot creation time as new ceph.snap.btime vxattr

ddiss opened the pull request:
Samba requires the snapshot creation time for the purpose of tracking and identifying Previous Versions for files and directories.

Fixes: https://tracker.ceph.com/issues/38838

gregsfortytwo commented on src/test/libcephfs/test.cc:
```diff
@@ -2185,3 +2185,47 @@ TEST(LibCephFS, OperationsOnDotDot) {
+  ASSERT_EQ('.', *s);
+
+  // file within the snapshot should carry the same btime
+  sprintf(c_temp, "/.snap/test_snap_xattr_snap_%d/%s", mypid, test_snap_xattr_file);
```
I know it's a bit trivial and silly in black-box terms, but I'd feel better if we also make sure that creating a second snapshot and checking its time results in a different time that is newer than the first one as a bit of a white-box test resilient against future changes.

ddiss replied:
Sounds good. I'll add this functionality. Thanks for the feedback!

ddiss commented:
The updated test compares the ceph.snap.btime between an older and newer snapshot.

batrick commented on src/test/libcephfs/test.cc:
```diff
@@ -2185,3 +2185,67 @@ TEST(LibCephFS, OperationsOnDotDot) {
+  ASSERT_LT(0, alen);
+  ASSERT_LT(alen, xbuflen);
+  gxattrv[alen] = '\0';
+  char *s = strchrnul(gxattrv, '.');
```
instead:

````
utime_t btime;
if (!btime.parse(gxattrv)) {
  ASSERT_EQ(0, 1);
}
````

batrick commented on src/test/libcephfs/test.cc:
```diff
@@ -2185,3 +2185,67 @@ TEST(LibCephFS, OperationsOnDotDot) {
+  alen = ceph_getxattr(cmount, c_temp, "ceph.snap.btime", (void *)gxattrv2, xbuflen);
+  ASSERT_LT(0, alen);
+  ASSERT_LT(alen, xbuflen);
+  gxattrv2[alen] = '\0';
```
ditto

ddiss commented on src/client/Client.cc:
```diff
@@ -11791,6 +11792,18 @@ size_t Client::_vxattrcb_dir_pin(Inode *in, char *val, size_t size)
+{
+  return snprintf(val, size, "%llu.09%lu",
+      (long long unsigned)in->snap_btime.sec(),
+      (long unsigned)in->snap_btime.nsec());
```
This adds a zero termination to the xattr value (which is included in the returned length), I'll assume that's desired in this case.

batrick replied:
? Both uses of snprintf add a NUL character? `snprintf` does not include the NUL character in the return count.

ddiss replied:
I'd prefer not to go down the [u]timespan format route for a few reasons:
- it's not consistent with the existing rctime vxattr format (which is also secs.nsecs)
- it's harder to parse on the client side
  - currently the only consumer is Samba,  which does a simple struct timespec conversion

ddiss commented on src/mds/CInode.cc:
```diff
@@ -3331,7 +3331,20 @@ int CInode::encode_inodestat(bufferlist& bl, Session *session,
+    realm->get_snap_info(infomap,
+                         snapid,  // min
+                         snapid); // max
+    if (infomap.size() == 1) {
```
Fixed

ddiss commented:
New version adds Fixes: https://tracker.ceph.com/issues/38838 tags and reworks the infomap.size() == 1 logic.

ukernel commented on src/client/Client.cc:
```diff
@@ -11888,6 +11909,14 @@ const Client::VXattr Client::_file_vxattrs[] = {
+    hidden: false,
+    exists_cb: &Client::_vxattrcb_snap_btime_exists,
+    flags: 0,
+  },
```
_vxattrcb_snap_btime_exists () should always return false for file

ukernel commented:
LGTM

ddiss commented:
Thanks for the feedback so far - the corresponding kernel changes were proposed (and subsequently acked) via https://www.spinics.net/lists/ceph-devel/msg44894.html .

@batrick any further comment on the proposed $secs.$nsecs format?

ddiss commented:
I've posted the Samba changes to https://lists.samba.org/archive/samba-technical/2019-March/133089.html - would be good to get final feedback on the time format soon if possible, as that appears to be the only thing holding up all three patch-sets.

batrick commented:
Investigating a possible QA failure caused by this PR:

http://pulpito.ceph.com/pdonnell-2019-04-02_00:10:11-fs-master-distro-basic-smithi/
http://pulpito.ceph.com/pdonnell-2019-04-02_00:13:00-fs-wip-pdonnell-testing-20190401.220144-distro-basic-smithi/

batrick commented on src/test/libcephfs/test.cc:
```diff
@@ -2185,3 +2185,67 @@ TEST(LibCephFS, OperationsOnDotDot) {
+  *s = '\0';
+  utime_t new_btime = utime_t(strtoull(gxattrv2, NULL, 10), strtoull(s + 1, NULL, 10));
+  ASSERT_LT(btime, new_btime);
+
```
So we are getting these failures now:

http://pulpito.ceph.com/pdonnell-2019-04-02_00:13:00-fs-wip-pdonnell-testing-20190401.220144-distro-basic-smithi/

I guess this new xattr must be the cause but I don't see any tests doing snapshots before the unit test that fails. Can you try this out via `ceph_test_libcephfs` to see if you can reproduce locally?

ddiss replied:
I've been able to reproduce this failure locally and will look into it - thanks for the heads-up.

ddiss commented:
I think I've finally worked out what's going on here:

Client::_listxattr() returns a length based on the static _vxattrs_name_size() value. Client::_vxattrs_calcu_name_size() only takes into account whether the vxattr is hidden, **not** whether the vxattr.exists_cb() returns true/false. 
When filling the xattr buffer, Client::_listxattr() checks vxattr.hidden **and** vxattr.exists_cb(). When the latter returns false (as is the case for ceph.snap.btime on non-snapshots), we return a length which is larger than the amount that we wrote to the buffer.

ddiss commented:
It seems that "ceph.snap.btime" is the only vxattr to set hidden=false and fill the exists_cb(), so I don't think this client bug affects existing releases.
IMO there are a couple of options to fix this:

1. drop _dir_vxattrs_name_size / _file_vxattrs_name_size and always calculate these lengths based on whether exists_cb() returns true in the Client::_listxattr()  code-path.
2. flag the "ceph.snap.btime" vxattr as hidden and add some sort of assert to ensure that hidden is always set to true alongside an exists_cb() callback.

I'll proceed with option (1), but any other feedback would still be appreciated.

ddiss commented:
Changes since previous version:

- rebase against master
- add four new commits (on top) which:
  - fix Client::_listxattr() vxattr length calculations
  - extend listxattr test coverage

batrick commented on src/client/Client.cc:
```diff
@@ -11342,44 +11340,55 @@ int Client::ll_getxattr(Inode *in, const char *name, void *value,
+  r = 0;
+  for (map<string,bufferptr>::iterator p = in->xattrs.begin();
+       p != in->xattrs.end();
+       ++p) {
```
As a rule in new code, when changing or creating new loops like this, we use the generic structured for loop. i.e.:

`for (const auto& p : in->xattrs)`

batrick commented on src/client/Client.cc:
```diff
@@ -11342,44 +11340,55 @@ int Client::ll_getxattr(Inode *in, const char *name, void *value,
+  for (map<string,bufferptr>::iterator p = in->xattrs.begin();
+       p != in->xattrs.end();
+       ++p) {
+    this_len = p->first.length() + 1;
```
Please move `this_len` declaration here.

batrick commented on src/client/Client.cc:
```diff
@@ -11342,44 +11340,55 @@ int Client::ll_getxattr(Inode *in, const char *name, void *value,
+    if (vxattr->exists_cb && !(this->*(vxattr->exists_cb))(in))
+      continue;
+
+    this_len = vxattr->name.length() + 1;
```
... and redeclare here.

ddiss commented:
Changes since previous version:

- address Patrick's "fix _listxattr() vxattr buffer length calculation" comments
- rebase against master again

ddiss commented:
> Client::_listxattr() returns a length based on the static _vxattrs_name_size() value. Client::_vxattrs_calcu_name_size() only takes into account whether the vxattr is hidden, **not** whether the vxattr.exists_cb() returns true/false.
> When filling the xattr buffer, Client::_listxattr() checks vxattr.hidden **and** vxattr.exists_cb(). When the latter returns false (as is the case for ceph.snap.btime on non-snapshots), we return a length which is larger than the amount that we wrote to the buffer.

FWIW, it looks like the kernel ceph_listxattr() function has similar behaviour, so I'll respin my kernel patchset with a corresponding fix once this gets merged.
